Classify horizontal tab as a blank character

M-Moawad opened the pull request:
Update the `_CTYPE_DATA_0_127` macro to include `_B` flag for the horizontal tab character (0x09), ensuring it is correctly recognized as a blank character.  

This change ensures that both horizontal tab and space characters are appropriately classified as blank character.

keith-packard commented:
Alas, this breaks `isprint`, which is not supposed to return true for tab. Picolibc has a special macro for `isblank` which includes an explicit test for tab along with the bitmask test for `_B`. This breaks C++, as discussed previously here: https://github.com/picolibc/picolibc/issues/778 I think the only correct fix is to expand the table to 16 bits so that there are unique bits for both `isblank` and `isprint` instead of trying to re-use the blank bit for both.

M-Moawad commented:
@keith-packard 
Could you please review the last commit for the addition of the _T macro? Is that approach acceptable, or would you suggest a different solution?

keith-packard commented:
The C++ standard requires masks for each of:
```
static const mask space = 1 << 0;
static const mask print = 1 << 1;
static const mask cntrl = 1 << 2;
static const mask upper = 1 << 3;
static const mask lower = 1 << 4;
static const mask alpha = 1 << 5;
static const mask digit = 1 << 6;
static const mask punct = 1 << 7;
static const mask xdigit = 1 << 8;
static const mask blank = 1 << 9;
static const mask alnum = alpha | digit;
static const mask graph = alnum | punct;
```
We've currently got these:
```
#define	_U	01
#define	_L	02
#define	_N	04
#define	_S	010
#define _P	020
#define _C	040
#define _X	0100
#define	_B	0200
```
 * `space` is `_S`.
 * `print` is `_P|_U|_L|_N|_B`
 * `cntrl` is `_C`.
 * `upper` is `_U`
 * `lower` is the same as `_L`
 * `alpha` is `_U|_L`
 * `digit` is `_N`
 * `punct` is `_P`
 * `xdigit` is `_X|_N`
 * `blank` is `_B` or `c == '\t'`.
I think it's probably OK for our masks to include more than one bit; the C++ spec doesn't require any particular value for them. So, your plan of adding a _T value which maps only to `'\t'` seems like it should work fine. We'll have to create a patch for gcc/libstdc++ that adds picolibc-specific ctype bits, finally creating a picolibc configuration for that code. That's gonna be a bigger lift than any of the picolibc changes, alas. I'm not sure how llvm's libc++ manages this, but we'll need to go see how to make that work too.

Because of the C++ library issue, we'll need to think about a migration strategy which allows existing C++ library bits to continue to work while allowing new C++ library bits to get the bug fix. I think what that will require is for us to continue to expose `_ctype_` in its current form so that C++ libraries can use their existing implementation of `ctype<char>::classic_table()`. We would then copy the `_ctype_` bits into the new wider table with a new name and have that one include your new `_T` bit for `'\t'`. This way, we can ship picolibc with the added bits. Hrm. Now I'm questioning whether picolibc itself should use the wide table or the narrow table with the special `isblank` implementation. The only reason for this change is to fix the C++ libraries; there's no bug in the C library spec as we have the magic `isblank` code, right?

M-Moawad commented:
>```
> alas. I'm not sure how llvm's libc++ manages this, but we'll need to go see how to make that work too.

For LLVM's libc++, I checked the relevant code here: https://github.com/llvm/llvm-project/blob/main/libcxx/include/__locale#L389C1-L405C52

```c
#elif defined(_NEWLIB_VERSION)
  // Same type as Newlib's _ctype_ array in newlib/libc/include/ctype.h.
  typedef char mask;
  // In case char is signed, static_cast is needed to avoid warning on
  // positive value becomming negative.
  static const mask space  = static_cast<mask>(_S);
  static const mask print  = static_cast<mask>(_P | _U | _L | _N | _B);
  static const mask cntrl  = static_cast<mask>(_C);
  static const mask upper  = static_cast<mask>(_U);
  static const mask lower  = static_cast<mask>(_L);
  static const mask alpha  = static_cast<mask>(_U | _L);
  static const mask digit  = static_cast<mask>(_N);
  static const mask punct  = static_cast<mask>(_P);
  static const mask xdigit = static_cast<mask>(_X | _N);
  static const mask blank  = static_cast<mask>(_B);
  // mask is already fully saturated, use a different type in regex_type_traits.
  static const unsigned short __regex_word = 0x100;
```
After making the following changes, I tested LLVM's libc++, and it worked:
    -  Changed the `mask` type to `short`: `typedef short mask;`
    - Added **_T** to the `blank` mask:  `static const mask blank  = static_cast<mask>(_B | _T);`
    - Updated the **__regex_word** value: `static const unsigned short __regex_word = 0x200;`

>```
> We would then copy the _ctype_ bits into the new wider table with a new name and have that one include your new _T bit for '\t'.

Introducing a wider type for `_ctype_` would require changes in the libc++ implementations across all toolchains using Picolibc. 
Before proceeding, how many toolchains currently use Picolibc and determine if they are open source. If the number is manageable, we could consider submitting PRs to update their libc++ implementations to accommodate this change. 

>```
>The only reason for this change is to fix the C++ libraries; there's no bug in the C library spec as we have the magic isblank code, right?

Yes, that is correct. this to fix C++ libraries only, and the Picolibc itself doesn't depend on this table, so no bug in the C library spec.

M-Moawad commented:
After checking the Picolibc locale, I found that it depends on narrow table with char type, so I will add another commit to add the wide table and use this table only with Libc++ to fix the issue of `isblank`

keith-packard commented:
> After making the following changes, I tested LLVM's libc++, and it worked: - Changed the `mask` type to `short`: `typedef short mask;` - Added **_T** to the `blank` mask: `static const mask blank = static_cast<mask>(_B | _T);` - Updated the **__regex_word** value: `static const unsigned short __regex_word = 0x200;`

Awesome. We'll obviously need to move those to a picolibc-specific block in libc++ so that newlib continues to operate.

> Introducing a wider type for `_ctype_` would require changes in the libc++ implementations across all toolchains using Picolibc. Before proceeding, how many toolchains currently use Picolibc and determine if they are open source. If the number is manageable, we could consider submitting PRs to update their libc++ implementations to accommodate this change.

Yeah, that's why I suggested leaving `_ctype_` alone and creating a new table with the wide bits. That's easy for the C locale, but I haven't looked at how hard it would be for other locales.

> For using wide table or narrow table, let me check this because Picolibc locale depends on narrow table and when using short int table (wide table) instead of char table (narrow table), I had issues for type incompatibility.

Maybe we don't care for the C locale where the memory impact is small?

keith-packard reviewed (approved):
This looks perfect.

keith-packard commented on newlib/libc/ctype/ctype_.c:
```diff
@@ -98,6 +109,9 @@ const char _ctype_[1 + 256] = {
 
 #endif	/* !ALLOW_NEGATIVE_CTYPE_INDEX */
 
+const char _ctype_wide[1 + 256] = { 0, _CTYPE_DATA_0_127_WIDE,
```
this needs to be short instead of char.

M-Moawad replied:
@keith-packard Done

keith-packard commented:
I amended the commit message.
